### PR TITLE
Document query parameters directly instead of via freeform text blurbs

### LINKED ISSUE #665: Convert Swagger tag metadata (freeform blurbs) into parameter annotations
I volunteered to attempt to transfer the endpoint documentation currently implemented as freeform blurbs for "sections" (i.e. tags) in Swagger UI; into endpoint parameter-specific annotations.

Today, the freeform blurbs are stored in a list variable named `tags_metadata` in the file `nmdc_runtime/api/main.py`:

https://github.com/microbiomedata/nmdc-runtime/blob/ac6cf8a76ed54ed9c1f1c303d1b1726e0c9016ad/nmdc_runtime/api/main.py#L71-L292

An example of "parameter-specific annotations" is shown in this issue: https://github.com/microbiomedata/nmdc-runtime/issues/651

### PR DESCRIPTION
**Pull request by eecavanna**
In this branch, I effectively relocated the parameter-level documentation from the freeform blurbs atop two endpoint groups on Swagger UI, to the endpoints themselves in Swagger UI. That applies to the endpoints in the "find" section of the Swagger UI, and the endpoints whose paths begin with `/nmdcschema/` in the "metadata" section of Swagger UI.

### Details

The main challenge here was figuring out how to properly mix and match FastAPI's and Pydantic's abstractions in order to get descriptions of Pydantic model attributes to appear next to the corresponding parameter fields on the Swagger UI page. There seem to me to be some unresolved [upstream issues](https://github.com/fastapi/fastapi/discussions/8348) at the intersection of those three things:

- Encapsulating **query parameters** within a Pydantic model
- Displaying query parameter **descriptions** on Swagger UI
- FastAPI's integrations with Pydantic and Swagger/OpenAPI

image

In this PR branch, I landed on an approach that works for the `description` field, specifically. It does not work for the `title` and `examples` fields. In my local experimentation, I was able to get something to work for _all_ fields, but it involved converting the Pydantic models into regular Python classes (that _don't_ inherit from Pydantic's `BaseModel` class) and that was more of a deviation from the current state of things than I wanted to do in a documentation-focused PR.

I also added lots of `TODO` comments about adding documentation to functions that lack it (I just did so for functions I dealt with while working on this PR). `<soapbox>`I think a given piece of code is generally read more times than it is written (i.e. >1 versus 1), spending time deciphering code can take time/energy, and I think the person writing the code is generally best positioned to document their expectations about whatever it is they are implementing`</soapbox>`, so I plan to create a ticket(s) about the ones of these I think will be most impactful and assign them to the author(s) of those pieces of code.

Finally, here's a screenshot showing query parameter documentation appearing next to the parameters being described. Previously, those query parameters were documented in a freeform text blurb above that group of "metadata" endpoints. The difference shown in this screenshot is representative of the way documentation has changed for **many** endpoints in this PR branch.

image

### Related issue(s)

Fixes #665

### Related subsystem(s)

- [x] Runtime API (except the Minter)
- [ ] Minter
- [ ] Dagster
- [ ] Project documentation (in the `docs` directory)
- [ ] Translators (metadata ingest pipelines)
- [ ] MongoDB migrations
- [ ] Other

### Testing

- [x] I tested these changes (explain below)
- [ ] I did not test these changes

I confirmed all the tests performed by the GitHub Actions workflows still pass.

### Documentation

- [x] I **have not checked** for relevant documentation yet (e.g. in the `docs` directory)
- [ ] I have **updated** all relevant documentation so it will remain accurate
- [ ] Other (explain below)

### Maintainability

- [x] Every Python function I defined includes a docstring _(test functions are exempt from this)_
- [x] Every Python function parameter I introduced includes a type hint (e.g. `study_id: str`)
- [x] All "to do" or "fix me" Python comments I added begin with either `# TODO` or `# FIXME`
- [x] I used `black` to format all the Python files I created/modified
- [x] The PR title is in the imperative mood (e.g. "Do X") and not the declarative mood (e.g. "Does X" or "Did X")

### CODE REVIEW

**Comment by eecavanna** [comment]
I will mark this as "Ready for review" only so that the GitHub Actions workflows run on it; in particular, the one that runs the automated tests. I do not consider this branch to be ready for review, in the sense that I would merge it if someone approved it; but I do welcome reviews and feedback by anyone.

**Comment by eecavanna** [comment]
I've gone as far as I, independently, can with the endpoints defined in the files marked ✔️ below (this is a living list):
- [x] `api/endpoints/find.py`
- [ ] ...

**Comment by eecavanna** [comment]
I think I'll narrow the scope of the PR to be about only the "find" category of endpoints. I don't want to go through this process for the other endpoints this sprint. I have set up a model here that others (including my future self) can follow in order to get this done for all other endpoints.

**Comment by eecavanna** [comment]
Hi @dwinston, in case you review this PR branch and have time to address any of the `TODO`s I added, I'm OK with you making commits directly to this branch.

File: requirements/main.in [hunk 1]
```diff
@@ -9,7 +9,7 @@ dagster-graphql
 dagster-postgres
 dependency-injector
 dotted-dict
-fastapi>=0.104.1 # Pins Swagger UI version to 5.9.0 temporarily to handle a bug crashing it in 5.9.1
+fastapi>=0.115.0  # note: FastAPI 0.115.0 introduced support for encapsulating request _query_ parameters in Pydantic models, including Swagger annotations
```
Comment (eecavanna) [inline requirements/main.in:12]: Note: The `requirements.main.txt` file already lists this version of `fastapi`.

File: nmdc_runtime/api/models/util.py [hunk 2]
```diff
@@ -65,8 +176,21 @@ def set_page_if_cursor_unset(cls, values):
 
 
 class PipelineFindRequest(BaseModel):
-    pipeline_spec: str
-    description: str
+    """
+    TODO: Document this class and its attributes.
```
Comment (eecavanna) [inline nmdc_runtime/api/models/util.py:180]: Hi @dwinston, if you write a title (alias), description, and example value(s) for each of these two parameters, I'll take care of wrapping them in the necessary FastAPI/Pydantic stuff to get them to appear on the Swagger UI page.

File: nmdc_runtime/api/endpoints/metadata.py [hunk 3]
```diff
@@ -113,9 +123,26 @@ async def submit_changesheet(
...
+        Path(
+            title="Metadata file ObjectId",
+            description="The ObjectId (`_id`) of the metadata file you want to get.\n\n_Example_: `507f1f77bcf86cd799439011`",
+            examples=["507f1f77bcf86cd799439011"],
+        ),
+    ],
     mdb: MongoDatabase = Depends(get_mongo_db),
 ):
+    r"""
+    Undocumented
```
Comment (eecavanna) [inline nmdc_runtime/api/endpoints/metadata.py:137]: Hi @dwinston, will you summarize what this API endpoint was designed to do? I don't have any experience with GridFS and don't know how our users would get an ObjectId for a GridFS file (which I think is what the `object_id` parameter was designed to contain).

**Comment by eecavanna** [comment]
According to the [ticket](https://github.com/microbiomedata/issues/issues/947) for the next NMDC Release, code destined for that release is due in `main` by December 9. I would like for these changes to be included in that release.

**Comment by eecavanna** [comment]
I'm planning to merge this in on Thursday afternoon. Monday is the date on which we'll be freezing code so team members can begin testing ahead of the upcoming release.

File: nmdc_runtime/api/models/util.py (hunk 2 above)
Comment (dwinston) [reply nmdc_runtime/api/models/util.py]: see (commit message for) 442d6d420ef3e2bf675bfbcbfe34ba3e8ed693d4

File: nmdc_runtime/api/endpoints/metadata.py (hunk 3 above)
Comment (dwinston) [reply nmdc_runtime/api/endpoints/metadata.py]: done via e00c4bb11b1869041e08e4dd5f751680bf79fb1a

**Comment by eecavanna** [comment]
Thank you, @dwinston, for addressing so many of the `TODO`s. I think this will all make maintenance of the code base easier for people over time.

**Comment by eecavanna** [comment]
Thanks again, @dwinston. Merging into `main` now, ahead of Monday's code freeze.

**Pull request merged**